Junimatic/FishPondMachine.cs

```diff
@@ -16,37 +18,58 @@ public FishPondMachine(FishPond fishPond, Point accessPoint)
 
         public new FishPond Building => (FishPond)base.Building;
 
-        public override bool IsIdle => false;
+        public override MachineState State
+            => this.Building.output.Value is null ? MachineState.Working : MachineState.AwaitingPickup;
 
         public override bool FillMachineFromChest(GameStorage storage, Func<Item,bool> isShinyTest)
         {
-            // IsIdle being hard-coded to false should prevent this from being called.
+            // State never being Idle should prevent this from ever being called
             throw new NotImplementedException();
         }
 
-        public override bool FillMachineFromInventory(Inventory inventory)
+        public override void FillMachineFromInventory(Inventory inventory)
         {
-            // IsIdle being hard-coded to false should prevent this from being called.
+            // State never being Idle should prevent this from ever being called
             throw new NotImplementedException();
         }
 
         public override List<Item>? GetRecipeFromChest(GameStorage storage, Func<Item, bool> isShinyTest)
         {
-            // IsIdle being hard-coded to false should prevent this from being called.
+            // State never being Idle should prevent this from ever being called
             throw new NotImplementedException();
         }
 
         public override bool IsCompatibleWithJunimo(JunimoType projectType)
         {
             return projectType == JunimoType.Fishing;
         }
-        public override StardewValley.Object? HeldObject => this.Building.output.Value as StardewValley.Object;
-
-        protected override StardewValley.Object TakeItemFromMachine()
+        public override List<StardewValley.Item> GetProducts()
         {
             var oldValue = this.Building.output.Value;
             this.Building.output.Value = null;
-            return (StardewValley.Object)oldValue;
+            return [oldValue];
+        }
+
+        public override ProductCapacity CanHoldProducts(GameStorage storage)
+        {
+            var heldObject = this.Building.output.Value as StardewValley.Object;
+            if (heldObject is null)
+            {
+                throw new InvalidOperationException(I("CanHoldProducts should only be called if the State is AwaitingPickup"));
+            }
+
+            if (storage.CanAddToExistingStack(heldObject))
```

Comment:
This 'ProductCapacity' enum has a bad smell to me.  I can't quite say what I'd like better.  But right now you can point to:

1. CanAddToExistingStack and IsPossibleStorageFor both might iterate the contents of the chest, and you'd think that could be combined into a single iteration of the contents.
2. By "Existing Stack" that implies quality.  But if you have a chest with silver-quality grapes wouldn't that be the preferred chest for gold-quality grapes as well?
3. FishPondMachine only needs its own specialization because it has its own logic for getting the basis for deciding what similar objects are.

Now that I'm writing this I'm thinking maybe there should be something like an 'EstimateGetProduce" method that gives a maximum view of what the items might be so that the chest-picking logic can be better concentrated.